camp_2/buidl/protostar-buidl/src/voting.cairo

```diff
@@ -0,0 +1,228 @@
+    address: felt
+) -> (is_voter_registered: felt) {
+    let (registered) = registered_voter.read(address);
+    return (is_voter_registered=registered);
+}
+
+// ------
+// Constant Functions: non state-changing functions
+// ------
+
+// @dev Check if the caller is allowed to vote
+// @implicit syscall_ptr (felt*)
+// @implicit pedersen_ptr (HashBuiltin*)
+// @implicit range_check_ptr
+// @param info (VoterInfo): struct indicating whether the voter is allowed to vote
+func _assert_allowed{syscall_ptr: felt*, range_check_ptr}(info: VoterInfo) {
+    with_attr error_message("VoterInfo: Your address is not allowed to vote.") {
+        assert_not_zero(info.allowed);
+    }
+
+    return ();
+}
+
+// ------
+// Non-Constant Functions: state-changing functions
+// ------
+
+// @dev Internal function to prepare the list of voters
+// @implicit syscall_ptr (felt*)
+// @implicit pedersen_ptr (HashBuiltin*)
+// @implicit range_check_ptr
+// @param registered_addresses_len (felt): number of registered voters
+// @param registered_addresses (felt*): array with the addresses of registered voters
+func _register_voters{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    registered_addresses_len: felt, registered_addresses: felt*
+) {
+    // No more voters, recursion ends
+    if (registered_addresses_len == 0) {
+        return ();
+    }
+
+    // Assign the voter at address 'registered_addresses[registered_addresses_len - 1]' a VoterInfo struct
+    // indicating that they have not yet voted and can do so
+    let votante_info = VoterInfo(allowed=1);
+    registered_voter.write(registered_addresses[registered_addresses_len - 1], 1);
+    voter_info.write(registered_addresses[registered_addresses_len - 1], votante_info);
+
+    // Go to next voter, we use recursion
+    return _register_voters(registered_addresses_len - 1, registered_addresses);
+}
+
+// @dev Given a vote, it checks if the caller can vote and updates the status of the vote, and the status of the voter
+// @implicit syscall_ptr (felt*)
+// @implicit pedersen_ptr (HashBuiltin*)
+// @implicit range_check_ptr
+// @param vote (felt): vote 0 or 1
+@external
+func vote{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(vote: felt) -> () {
+    alloc_locals;
+    Pausable.assert_not_paused();
+
+    // Know if a voter has already voted and continue if they have not voted
+    let (caller) = get_caller_address();
+    let (info) = voter_info.read(caller);
+    _assert_allowed(info);
+
+    // Mark that the voter has already voted and update in the storage
+    let info_actualizada = VoterInfo(allowed=0);
+    voter_info.write(caller, info_actualizada);
+
+    // Update the vote count with the new vote
+    let (status) = voting_status.read();
+    local updated_voting_status: VoteCounting;
+    if (vote == 0) {
+        assert updated_voting_status.votes_no = status.votes_no + 1;
```

Comment:
I would use a library function for addition to protect against overflow

Reply:
After a conversation in Slack with @barretodavid, we agreed on delaying this.